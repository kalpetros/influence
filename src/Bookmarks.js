import React, { useState, useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export const Bookmarks = () => {
  const [tree, setTree] = useState([]);
  const [data, setData] = useState([]);

  useEffect(() => {
    setData(tree);
  }, [tree]);

  useEffect(() => {
    chrome.bookmarks.getTree(function (response) {
      setTree(response);
    });
  }, []);

  const handleClick = (e) => {
    const id = e.currentTarget.id;

    chrome.bookmarks.getSubTree(id, function (response) {
      setData(response);
    });
  };

  const handleBackClick = (e) => {
    chrome.bookmarks.getSubTree(id, function (response) {
      setData(response);
    });
  };

  const items = data.map((node) => {
    return node.children.map((child, index) => {
      if (typeof child.url !== 'undefined') {
        return (
          <a
            key={`bookmark-${child.id}-${index}`}
            className="p-5 border-b block cursor-pointer text-gray-700"
            href={child.url}
            target="__blank"
          >
            <img
              className="inline mr-5"
              src={`https://plus.google.com/_/favicon?domain_url=${child.url}`}
            />
            {child.title}
          </a>
        );
      }

      return (
        <div
          key={`bookmark-${child.id}-${index}`}
          id={child.id}
          className="p-5 border-b block cursor-pointer text-gray-700"
          onClick={handleClick}
        >
          <FontAwesomeIcon
            icon="folder"
            size="lg"
            className="mr-5 cursor-pointer"
          />
          {child.title}
        </div>
      );
    });
  });

  return (
    <div className="overflow-auto">
      <div className="p-5">{items}</div>
    </div>
  );
};
