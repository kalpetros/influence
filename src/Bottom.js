import React from 'react';
import PropTypes from 'prop-types';

import { TimeWidget } from './TimeWidget';
import { WeatherWidget } from './WeatherWidget';

export const Bottom = (props) => {
  const onWeatherClick = props.onWeatherClick;

  return (
    <div className="grid grid-cols-2 items-end p-5">
      <TimeWidget />
      <WeatherWidget onClick={onWeatherClick} />
    </div>
  );
};

Bottom.propTypes = {
  onWeatherClick: PropTypes.func.isRequired,
};
