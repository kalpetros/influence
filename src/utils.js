import React from 'react';
import moment from 'moment';

// Convert wind speed from m/s to beaufort scale
export function toBeaufort(wind) {
  let beaufort = 0;
  if (wind < 0.3) {
    beaufort = 0;
  } else if (wind >= 0.3 && wind <= 1.5) {
    beaufort = 1;
  } else if (wind >= 1.6 && wind <= 3.3) {
    beaufort = 2;
  } else if (wind >= 3.4 && wind <= 5.4) {
    beaufort = 3;
  } else if (wind >= 5.5 && wind <= 7.9) {
    beaufort = 4;
  } else if (wind >= 8 && wind <= 10.7) {
    beaufort = 5;
  } else if (wind >= 10.8 && wind <= 13.8) {
    beaufort = 6;
  } else if (wind >= 13.9 && wind <= 17.1) {
    beaufort = 7;
  } else if (wind >= 17.2 && wind <= 20.7) {
    beaufort = 8;
  } else if (wind >= 20.8 && wind <= 24.4) {
    beaufort = 9;
  } else if (wind >= 24.5 && wind <= 28.4) {
    beaufort = 10;
  } else if (wind >= 28.5 && wind <= 32.6) {
    beaufort = 11;
  } else if (wind >= 32.7) {
    beaufort = 12;
  }
  return beaufort;
}

export function formatBytes(bytes, decimals) {
  if (bytes == 0) return '0 Byte';
  const k = 1000; // or 1024 for binary
  const dm = decimals + 1 || 3;
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
}

export function getSettings() {
  const settings = localStorage.getItem('settings');

  if (settings === null) {
    const settings = {
      time: '24',
      temperature: 'celsius',
      pressure: 'pascal',
      wind: 'ms',
      darkMode: 'false',
    };

    localStorage.setItem('settings', JSON.stringify(settings));

    return settings;
  }

  return JSON.parse(settings);
}

export function getGreeting() {
  const date = moment();
  const hours = date.hours();

  if (hours >= 17 || hours < 5) {
    return 'Good evening,';
  }

  if (hours >= 12 || hours < 17) {
    return 'Good afternoon,';
  }

  return 'Good morning,';
}

export function formatTime(obj) {
  const settings = getSettings();
  let time = obj.format('HH:mm');

  if (settings.time === '12') {
    time = obj.format('h:mm');
  }

  return time;
}

export const formatTemperature = (temperature) => {
  const settings = getSettings();
  temperature = Math.round(parseFloat(temperature), 1);

  if (settings.temperature === 'fahrenheit') {
    temperature = Math.round(temperature * 1.8 + 32, 0);
    return (
      <>
        {temperature} <i className="wi wi-fahrenheit"></i>
      </>
    );
  }

  return (
    <>
      {temperature} <i className="wi wi-celsius"></i>
    </>
  );
};

export function formatPressure(pressure) {
  const settings = getSettings();
  pressure = Math.round(parseFloat(pressure), 0);

  if (settings.pressure === 'pascal') {
    // 1 hPa (hectopascal) = 1 millibar = 100 Pa
    pressure = pressure * 100;
    return `${pressure} Pa`;
  }

  return `${pressure} mbar`;
}

export const formatWind = (wind) => {
  const settings = getSettings();

  if (settings.wind === 'beaufort') {
    wind = toBeaufort(parseFloat(wind));
    return <i className={`wi wi-wind-beaufort-${wind}`}></i>;
  }

  return `${wind} m/s`;
};

export const DB = {
  name: 'InfluenceDB',
  storeName: 'todo',
  version: 1,
  open: () => {
    return new Promise((resolve, reject) => {
      let db;

      const request = window.indexedDB.open(DB.name, DB.version);

      request.onsuccess = (e) => {
        db = e.target.result;

        resolve('success');

        db.onsuccess = (e) => {
          resolve('success');
        };

        db.onerror = (e) => {
          reject(e.target.errorCode);
        };
      };

      request.onerror = (e) => {
        reject(e.target.errorCode);
      };

      request.onupgradeneeded = (e) => {
        const db = e.target.result;
        const store = db.createObjectStore(DB.storeName, {
          autoIncrement: true,
        });

        store.createIndex('date', 'date', { unique: false });

        resolve('success');
      };
    });
  },
  add: (item) => {
    return new Promise((resolve, reject) => {
      let db;

      const request = window.indexedDB.open(DB.name, DB.version);

      request.onsuccess = (e) => {
        db = e.target.result;

        db.onsuccess = (e) => {
          resolve('success');
        };

        db.onerror = (e) => {
          reject(e.target.errorCode);
        };

        const transaction = db.transaction(DB.storeName, 'readwrite');
        const store = transaction.objectStore(DB.storeName);
        store.add(item);
        resolve('success');
      };

      request.onerror = (e) => {
        reject(e.target.errorCode);
      };
    });
  },
  put: (item, key) => {
    return new Promise((resolve, reject) => {
      let db;

      if (isNaN(key)) {
        reject('Key should be an integer');
        return;
      }

      const request = window.indexedDB.open(DB.name, DB.version);

      request.onsuccess = (e) => {
        db = e.target.result;

        db.onsuccess = (e) => {
          resolve('success');
        };

        db.onerror = (e) => {
          reject(e.target.errorCode);
        };

        const transaction = db.transaction(DB.storeName, 'readwrite');
        const store = transaction.objectStore(DB.storeName);
        const request = store.put(item, key);

        request.onsuccess = (e) => {
          resolve('success');
        };
      };

      request.onerror = (e) => {
        reject(e.target.errorCode);
      };
    });
  },
  getKeys: () => {
    return new Promise((resolve, reject) => {
      let db;

      const request = window.indexedDB.open(DB.name, DB.version);

      request.onsuccess = (e) => {
        db = e.target.result;

        db.onsuccess = (e) => {
          resolve('success');
        };

        db.onerror = (e) => {
          reject(e.target.errorCode);
        };

        const transaction = db.transaction(DB.storeName, 'readonly');
        const store = transaction.objectStore(DB.storeName);
        const request = store.getAllKeys();

        request.onsuccess = (e) => {
          const keys = e.target.result;
          resolve(keys);
        };
      };

      request.onerror = (e) => {
        reject(e.target.errorCode);
      };
    });
  },
  get: (key) => {
    return new Promise((resolve, reject) => {
      let db;

      const request = window.indexedDB.open(DB.name, DB.version);

      request.onsuccess = (e) => {
        db = e.target.result;

        db.onsuccess = (e) => {
          resolve('success');
        };

        db.onerror = (e) => {
          reject(e.target.errorCode);
        };

        const transaction = db.transaction(DB.storeName, 'readonly');
        const store = transaction.objectStore(DB.storeName);
        const request = store.get(key);

        request.onsuccess = (e) => {
          const result = e.target.result;
          result['key'] = key;
          resolve(result);
        };
      };

      request.onerror = (e) => {
        reject(e.target.errorCode);
      };
    });
  },
};
