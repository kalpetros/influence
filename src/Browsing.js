import React, { useState, useContext } from 'react';

import { Modal } from './Modal';
import { RecentBookmarks } from './RecentBookmarks';
import { Bookmarks } from './Bookmarks';
import { History } from './History';
import { TopVisited } from './TopVisited';
import { Downloads } from './Downloads';
import { Todo } from './Todo';
import { ThemeContext } from './store';

export const Browsing = () => {
  const [view, setView] = useState('topVisited');
  const { state: theme } = useContext(ThemeContext);

  const handleClick = (e) => {
    setView(e.currentTarget.id);
  };

  const page =
    view === 'topVisited' ? (
      <TopVisited />
    ) : view === 'recentBookmarks' ? (
      <RecentBookmarks />
    ) : view === 'bookmarks' ? (
      <Bookmarks />
    ) : view === 'history' ? (
      <History />
    ) : view === 'downloads' ? (
      <Downloads />
    ) : view === 'todo' ? (
      <Todo />
    ) : null;

  const menuItems = [
    {
      id: 'topVisited',
      name: 'Top Visited',
    },
    {
      id: 'recentBookmarks',
      name: 'Recent Bookmarks',
    },
    {
      id: 'bookmarks',
      name: 'Bookmarks',
    },
    {
      id: 'history',
      name: 'History',
    },
    {
      id: 'downloads',
      name: 'Downloads',
    },
    {
      id: 'todo',
      name: 'Todo',
    },
  ];

  return (
    <Modal
      icon="bars"
      iconSize="2x"
      iconColor={theme.menuIconColor}
      name="browsing"
      view={view}
      menuItems={menuItems}
      onClick={handleClick}
    >
      {page}
    </Modal>
  );
};
