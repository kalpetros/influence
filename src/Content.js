import React, { useState, useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { RecentBookmarks } from './RecentBookmarks';
import { Bookmarks } from './Bookmarks';
import { History } from './History';
import { TopVisited } from './TopVisited';
import { Downloads } from './Downloads';
import { Top } from './Top';

const Menu = (props) => {
  const { view: view, onClick: onClick, onClose: onClose } = props;
  const items = [
    {
      id: 'topVisited',
      name: 'Top Visited',
    },
    {
      id: 'recentBookmarks',
      name: 'Recent Bookmarks',
    },
    {
      id: 'bookmarks',
      name: 'Bookmarks',
    },
    {
      id: 'history',
      name: 'History',
    },
    {
      id: 'downloads',
      name: 'Downloads',
    },
  ];

  const tabs = items.map((item) => {
    let baseClass = 'mr-1 bg-white inline-block py-2 px-4 cursor-pointer';
    let className = `${baseClass} text-blue-500`;

    if (view === item.id) {
      className = `${baseClass} -mb-px border-l border-t border-r rounded-t text-blue-700 font-semibold`;
    }

    return (
      <li key={item.id} id={item.id} className={className} onClick={onClick}>
        {item.name}
      </li>
    );
  });

  return (
    <div className="sticky top-0 bg-white">
      <div className="p-5 text-right">
        <FontAwesomeIcon
          icon="times"
          size="lg"
          className="cursor-pointer"
          onClick={onClose}
        />
      </div>
      <div>
        <ul className="flex border-b">{tabs}</ul>
      </div>
    </div>
  );
};

export const Content = () => {
  const [view, setView] = useState('topVisited');

  const handleClick = (e) => {
    setView(e.currentTarget.id);
  };

  const handleClose = () => {
    alert('adasd');
  };

  const page =
    view === 'topVisited' ? (
      <TopVisited />
    ) : view === 'recentBookmarks' ? (
      <RecentBookmarks />
    ) : view === 'bookmarks' ? (
      <Bookmarks />
    ) : view === 'history' ? (
      <History />
    ) : view === 'downloads' ? (
      <Downloads />
    ) : null;

  return (
    <div className="fixed top-0 w-2/3 h-screen overflow-auto bg-white">
      <Menu view={view} onClick={handleClick} onClose={handleClose} />
      {page}
    </div>
  );
};
