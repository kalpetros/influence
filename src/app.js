import React, { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';

import { Layout } from './Layout';
import { Top } from './Top';
import { TodoWidget } from './TodoWidget';
import { Bottom } from './Bottom';
import { SettingsStateProvider } from './store';

import { library } from '@fortawesome/fontawesome-svg-core';
import { fab } from '@fortawesome/free-brands-svg-icons';
import {
  faBookmark,
  faHistory,
  faList,
  faDownload,
  faBars,
  faCog,
  faTimes,
  faFolder,
  faCheck,
  faExclamationTriangle,
  faSearch,
  faChevronLeft,
  faInfoCircle,
  faSpinner,
  faPause,
} from '@fortawesome/free-solid-svg-icons';

library.add(
  fab,
  faBookmark,
  faHistory,
  faList,
  faDownload,
  faBars,
  faCog,
  faTimes,
  faFolder,
  faCheck,
  faExclamationTriangle,
  faSearch,
  faChevronLeft,
  faInfoCircle,
  faSpinner,
  faPause
);

const App = () => {
  return (
    <SettingsStateProvider>
      <Layout>
        <Top />
        <TodoWidget />
        <Bottom />
      </Layout>
    </SettingsStateProvider>
  );
};

ReactDOM.render(<App />, document.getElementById('root'));
