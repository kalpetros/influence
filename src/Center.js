import React, { useState, useEffect } from 'react';
import moment from 'moment';

import { Datepicker } from './Datepicker';
import { DB } from './utils';
import { getGreeting } from './utils';

export const Center = () => {
  const dateNow = moment().toISOString();
  const [data, setData] = useState([]);
  const [todo, setTodo] = useState('');
  const [date, setDate] = useState(dateNow);

  useEffect(() => {
    fetchInit();
  }, []);

  const fetchInit = () => {
    DB.getKeys()
      .then((response) => {
        fetchData(response);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const fetchData = (keys) => {
    const promises = keys.map((key) => {
      return DB.get(key);
    });

    Promise.all(promises)
      .then((response) => {
        setData(response);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const handleInputChange = (e) => {
    setTodo(e.currentTarget.value);
  };

  const handleDateChange = (date) => {
    setDate(date);
  };

  const handleTodoClick = (e) => {
    const key = parseInt(e.currentTarget.dataset.key);

    let item = data.find((d) => {
      return parseInt(d.key) === key;
    });

    delete item['key'];
    item['status'] = item['status'] === 'todo' ? 'done' : 'todo';

    DB.put(item, key)
      .then((response) => {
        fetchInit();
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const handleSubmit = (e) => {
    if (e.key === 'Enter') {
      const item = {
        title: todo,
        date: date,
        status: 'todo',
      };

      DB.add(item)
        .then((response) => {
          setTodo('');
          fetchInit();
        })
        .catch((error) => {
          console.log(error);
        });
    }
  };

  const handleTodoDateChange = (date, todo) => {
    const key = parseInt(todo.key);

    let item = data.find((d) => {
      return parseInt(d.key) === key;
    });

    delete item['key'];
    item['date'] = date;

    DB.put(item, key)
      .then((response) => {
        fetchInit();
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const todos = data
    .sort((a, b) => {
      const aIsBeforeB = moment(a.date).isBefore(b.date);
      const bIsBeforeA = moment(b.date).isBefore(a.date);

      if (aIsBeforeB) {
        return -1;
      }

      if (bIsBeforeA) {
        return 1;
      }

      return 0;
    })
    .filter((todo) => {
      return todo.status === 'todo';
    })
    .map((todo, index) => {
      let titleClass = 'text-2xl text-gray-700';

      return (
        <li
          key={`todo-${index}`}
          className="grid grid-flow-col items-center py-5 pr-5 border-b border-gray-700 block cursor-pointer"
        >
          <div
            className={titleClass}
            data-key={todo.key}
            onClick={handleTodoClick}
          >
            {todo.title}
          </div>
          <div className="grid justify-end">
            <Datepicker
              isData
              date={todo.date}
              todo={todo}
              onChange={handleTodoDateChange}
            />
          </div>
        </li>
      );
    });

  return (
    <div className="grid max-content-rows-2 h-full w-8/12 m-auto overflow-hidden">
      <div>
        <h1 className="text-5xl">{getGreeting()}</h1>
        <div className="grid grid-flow-col items-center border-b border-gray-500">
          <div>
            <input
              className="appearance-none text-2xl bg-transparent h-20 w-full text-gray-700 leading-tight focus:outline-none"
              type="text"
              placeholder="What do you want to do?"
              value={todo}
              onChange={handleInputChange}
              onKeyDown={handleSubmit}
            />
          </div>
          <div className="grid justify-end pr-5">
            <Datepicker date={date} onChange={handleDateChange} />
          </div>
        </div>
      </div>
      <div className="mt-10 overflow-auto">
        <ul>{todos}</ul>
      </div>
    </div>
  );
};
