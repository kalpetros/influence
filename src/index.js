import React, { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';
import { Layout } from './Layout';
import { Content } from './Content';
import { Top } from './Top';
import { Center } from './Center';
import { Bottom } from './Bottom';

import { library } from '@fortawesome/fontawesome-svg-core';
import { fab, faAccusoft } from '@fortawesome/free-brands-svg-icons';
import {
  faBookmark,
  faHistory,
  faList,
  faDownload,
  faBars,
  faCog,
  faTimes,
  faFolder,
  faCheck,
  faExclamationTriangle,
} from '@fortawesome/free-solid-svg-icons';

library.add(
  fab,
  faBookmark,
  faHistory,
  faList,
  faDownload,
  faBars,
  faCog,
  faTimes,
  faFolder,
  faCheck,
  faExclamationTriangle
);

const App = () => {
  return (
    <Layout>
      <Content />
      <Top />
      <Center />
      <Bottom />
    </Layout>
  );
};

ReactDOM.render(<App />, document.getElementById('root'));
