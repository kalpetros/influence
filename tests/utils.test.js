import React from 'react';
import renderer from 'react-test-renderer';
import moment from 'moment';

import { toBeaufort } from '../src/utils';
import { formatBytes } from '../src/utils';
import { getSettings } from '../src/utils';
import { getGreeting } from '../src/utils';
import { formatTime } from '../src/utils';
import { formatTemperature } from '../src/utils';
import { formatPressure } from '../src/utils';
import { formatWind } from '../src/utils';

test('Test m/s to beaufort convertion', () => {
  const msUnit = 10;
  const beaufort = toBeaufort(msUnit);

  expect(beaufort).toBe(5);
});

test('Test 1000 bytes to KB convertion', () => {
  const bytes = 10;
  const result = formatBytes(bytes);

  expect(result).toBe('10 Bytes');
});

test('Test 1000000 bytes to MB convertion', () => {
  const bytes = 1000000;
  const result = formatBytes(bytes);

  expect(result).toBe('1 MB');
});

test('Test getting settings from local storage', () => {
  const settings = getSettings();

  expect(settings).toBeDefined();
});

test('Test settings number of keys', () => {
  const settings = getSettings();
  const nofKeys = Object.keys(settings).length;

  expect(nofKeys).toBe(5);
});

test('Test Greeting output', () => {
  const greeting = getGreeting();
  expect(greeting).toBe('Good evening,');
});

test('Test time formatting', () => {
  const time = moment('2020-05-05 12:00');
  const result = formatTime(time);
  expect(result).toBe('12:00');
});

test('Test time formatting with invalid date', () => {
  const time = moment('2020-20-12 12:00');
  const result = formatTime(time);
  expect(result).toBe('Invalid date');
});

test('Test temperature formatting', () => {
  const temperature = 25;
  const element = renderer.create(formatTemperature(temperature));
  const tree = element.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Test mbar to pascal pressure convertion', () => {
  const pressure = 1000;
  const result = formatPressure(pressure);
  expect(result).toBe('100000 Pa');
});

test('Test ms wind speed formatting', () => {
  const windSpeed = 200;
  const result = formatWind(windSpeed);
  expect(result).toBe('200 m/s');
});
