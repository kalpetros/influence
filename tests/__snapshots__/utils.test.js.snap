// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test temperature formatting 1`] = `
Array [
  "25",
  " ",
  <i
    className="wi wi-celsius"
  />,
]
`;
